Fix repeating token sentence boundary bug

kotman12 opened the pull request:
Fix sentence boundary detection bug in case of repeating tokens (i.e. while using OpenNLP analysis chain in conjunction with a KeywordRepeatFilter) by keeping track of the sentence index and looking ahead one token. Move inner sentence iteration to a component to be shared by the sentence-aware OpenNLP filters.

kotman12 committed:
add test to show how keyword repeat filter interferes with sentence boundary transition in downstream filters

kotman12 committed:
implement SentenceAttributeExtractor with one token lookahead to properly handle sentence boundary when running with KeywordRepeatFilter

kotman12 committed:
implement SentenceAttributeExtractor with one token lookahead to properly handle sentence boundary when running with KeywordRepeatFilter

kotman12 committed:
lemmaNum needs to be a valid index

kotman12 committed:
remove unnecessary member variables

kotman12 committed:
simplify readNextSentence condition

kotman12 committed:
refactor OpenNLPChunkerFilter to use SentenceAttributeExtractor

kotman12 committed:
remove unnecessary member from OpenNLPPOSFilter, refine SentenceAttribute

kotman12 committed:
remove unnecessary setting of end-of-sentence bit

kotman12 committed:
tidy opennlp

kotman12 committed:
tidy SentenceAttribute

kotman12 committed:
grammar

kotman12 committed:
rename areMoreTokensAvailable to areMoreSentencesAvailable for clarity

kotman12 committed:
more specific test input data name

Signed-off-by: lkotzaniewsk <lkotzaniewsk@bloomberg.net>

kotman12 commented:
Linking issue #11735

kotman12 committed:
add license boilerplate

Signed-off-by: lkotzaniewsk <lkotzaniewsk@bloomberg.net>

kotman12 committed:
tidy test

kotman12 committed:
tidy and add javadoc

kotman12 commented:
./gradlew check passed locally as described in the contribution guide 😃 
![image](https://user-images.githubusercontent.com/13710476/188029979-f0d271ce-5999-46fc-a585-aa3e6ae9c287.png)

kotman12 committed:
make exit and sentence extraction conditions more readable

kotman12 committed:
add test to catch early exit bug

kotman12 committed:
fix NPE during empty tokenStream

dweiss commented:
Hi @kotman12 . Sorry for the delay. I'm not that familiar with this part of the codebase but I think I see what's happening and how you managed to fix it. Looks good to me. It'd be good to run this pipeline over a larger text base to make sure there are no surprises and regressions.

dweiss commented on lucene/analysis/opennlp/src/java/org/apache/lucene/analysis/opennlp/SentenceAttributeExtractor.java:
```diff
@@ -0,0 +1,75 @@
+    do {
+      hasNextToken = input.incrementToken();
+      int currSentenceTmp = sentenceAtt.getSentenceIndex();
+      hasNext = currSentence == currSentenceTmp && hasNextToken;
```
```suggestion
      hasNext = (currSentence == currSentenceTmp && hasNextToken);
```
Makes it clearer to read.

dweiss commented on lucene/analysis/opennlp/src/test/org/apache/lucene/analysis/opennlp/TestOpenNLPLemmatizerFilterFactory.java:
```diff
@@ -290,4 +299,61 @@ public void testKeywordAttributeAwarenessDictionaryAndMaxEnt() throws Exception
+
+  public void testPreventEarlyExit() throws IOException {
+    ClasspathResourceLoader loader = new ClasspathResourceLoader(getClass());
+    InputStream earlyExitInput = loader.openResource("data/early-exit-bug-input.txt");
```
Can we add a try-with-finally around those opened resource streams?

kotman12 committed:
add outer parentheses for clarity

Co-authored-by: Dawid Weiss <dawid.weiss@gmail.com>

kotman12 committed:
wrap in try-resources block

kotman12 committed:
use try-with-finally instead

kotman12 commented:
> Hi @kotman12 . Sorry for the delay. I'm not that familiar with this part of the codebase but I think I see what's happening and how you managed to fix it. Looks good to me. It'd be good to run this pipeline over a larger text base to make sure there are no surprises and regressions.

Hi @dweiss  .. no worries, thanks for taking a look.

I have made the changes you suggested. I agree that it would be good to test this with a large and more varied text base. I did incorporate these changes separately as a patch in my current project which is how I found that my initial proposal actually had a regression (it threw an NPE in case of empty input). It's possible there other bugs that I haven't uncovered .. did you have any particular test dataset in mind?

-------------------------------------------------------------------------------------------------------------------------------------------
Totally separately I noticed that in the case of stacking two or more sentence-aware token filters, sentence token iteration and attribute cloning are performed separately across the different filters (this is the case in the current implementation as well). This work could probably be done in a dedicated `SentenceExtractionFilter` which could then pass a reference to a _single_ `sentenceTokenAttributes` to all downstream filters that rely on sentence-level analysis (perhaps such a reference could be stashed in the proposed `SentenceAttribute`). Also, following this reasoning a bit further, it is possible to conceive future implementations that rely on passing an arbitrary subset of adjacent tokens together to some analysis function, in which case the use of the term `Sentence` would become a misnomer. 

However, addressing these issues would be more effort so I wanted to check if it is even worth it to explore. I am starting to suspect that this package is not really used a lot because otherwise one would expect these bugs to have been caught sooner given the configuration is in the documentation.

dweiss commented:
> I am starting to suspect that this package is not really used a lot because otherwise one would expect these bugs to have been caught sooner given the configuration is in the documentation.

You're very like right. There's a lot of old cruft that's mostly unused lying around. Please add lucene/CHANGES.txt entry (under 9.5.0) and I think it's done.

kotman12 committed:
update CHANGES.txt + add comment

kotman12 commented:
@dweiss I updated CHANGES.txt but blew up this PR and messed up the history in the process. If you prefer this is a more concise PR with the relevant changes patched in -> https://github.com/apache/lucene/pull/11802

dweiss commented:
I don't know what happened there but I'm sure it's going to be fixable. Let me take a look later today or tomorrow morning (I'm out of office today).

dweiss commented:
I rebased your commits on top of main so that they're linear when merged. Waiting for builds to pass.

dweiss commented:
Thanks!

kotman12 commented:
Thanks as well for taking a look 👍
